### PR TITLE
Wrap Logging

### PR DESCRIPTION
**Pull request by quantumagi**
https://app.clickup.com/t/1nq21qq

This PR creates a standardized wrapper over everything related to logging - i.e.:

`using NLog` is removed and replaced with `using Stratis.Bitcoin.Configuration.Logging` where needed.

`this.logger = LogManager.GetCurrentClassLogger();` remains supported but so also the logger factory pattern - which I managed to get working correctly. The latter can be removed in a follow-up PR but in this PR it serves to keep the solution-wide changes minimal.

The interesting changes are found in the `Stratis.Bitcoin.Configuration.Logging` namespace and those mentioned in comments below. The overwhelming bulk of changes just relate to the `using` statements.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Dns.Tests/GivenADnsFeature.cs [hunk 1]
```diff
@@ -261,10 +261,12 @@ public void WhenDnsServerFailsToStart_ThenDnsFeatureRetries()
                 .Setup(f => f.Log(It.IsAny<LogLevel>(), It.IsAny<EventId>(), It.IsAny<It.IsAnyType>(), It.IsAny<Exception>(), (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()))
                 .Callback(new InvocationAction(invocation =>
                 {
-                    if (!serverError && (LogLevel)invocation.Arguments[0] == LogLevel.Error)
+                    if (!serverError && (LogLevel)invocation.Arguments[0] == LogLevel.Error && invocation.Arguments[4] is Func<object[], Exception, string> func)
```
Comment (zeptin) [inline src/Stratis.Bitcoin.Features.Dns.Tests/GivenADnsFeature.cs:264]: Are all these changes part of the logging PR?

Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Dns.Tests/GivenADnsFeature.cs:264]: Yes. These tests, for some reason, expect to find the log message in the string formatting parameters. This fixes that assumption by looking at the resultant formatted string instead.

**Comment by quantumagi** [comment]
Found two cases of malformed logging statements:

image

**Comment by quantumagi** [comment]
It seems Debug filtering is going to require another look before we can commit this.

**Comment by quantumagi** [comment]
This PR should be good now

File: src/Stratis.Bitcoin/Configuration/Logging/EventId.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Stratis.Bitcoin.Configuration.Logging
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/EventId.cs:1]: This code is commented out, should this be removed?

File: src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
+﻿using System;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Configuration.Logging
+{
+    /*
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs:7]: This code is commented out, should this be removed?

File: src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    public interface ILogger
+    {
+        void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter);
+
+        bool IsEnabled(LogLevel logLevel);
+    }
+    */
+    /*
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs:15]: This class too.

File: src/Stratis.Bitcoin/Configuration/Logging/ILoggerFactory.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿/*
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/ILoggerFactory.cs:1]: This code is commented out, should this be removed?

File: src/Stratis.Bitcoin/Configuration/Logging/LogLevel.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Stratis.Bitcoin.Configuration.Logging
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/LogLevel.cs:1]: This code is commented out, should this be removed?

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs [hunk 7]
```diff
@@ -29,36 +28,42 @@ public class ExtendedLoggerFactory : LoggerFactory
 
         public static ILoggerFactory Create()
         {
+            return new CustomLoggerFactory();
+            /*
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs:32]: This code is commented out, should this be removed?

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs [hunk 8]
```diff
@@ -29,36 +28,42 @@ public class ExtendedLoggerFactory : LoggerFactory
...
             }
             );
+            */
         }
 
         /// <summary>Loads the NLog.config file from the <see cref="DataFolder"/>, if it exists.</summary>
         public static ILoggerFactory Create(LogSettings settings, DataFolder dataFolder)
         {
+            return new CustomLoggerFactory();
+            /*
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs:49]: This code is commented out, should this be removed?

File: src/Stratis.Bitcoin/Configuration/Logging/NullLogger.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+
+namespace Stratis.Bitcoin.Configuration.Logging
+{
+    /*
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/NullLogger.cs:5]: This code is commented out, should this be removed?

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs [hunk 10]
```diff
@@ -131,17 +136,6 @@ static LoggingConfiguration()
             LogManager.ConfigurationReloaded += NLogConfigurationReloaded;
         }
 
-        /// <summary>Loads the NLog.config file from the <see cref="DataFolder"/>, if it exists.</summary>
-        public static void LoadNLogConfiguration(this ILoggerFactory loggerFactory, DataFolder dataFolder)
-        {
```
Comment (fassadlr) [inline src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs:136]: What happened to this logic Gustav?

**Comment by fassadlr** [comment]
@quantumagi did you test this branch with debug logging enabled? Does the other projects like Poa etc still do debug logging as they use the LogManager.GetCurrentClassLogger.

**Comment by fassadlr** [comment]
Also there are some tests still failing.

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs (hunk 10 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs:136]: This method was not being referenced.

File: src/Stratis.Bitcoin/Configuration/Logging/NullLogger.cs (hunk 9 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/NullLogger.cs:5]: Yes

File: src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs]: Removed.

File: src/Stratis.Bitcoin/Configuration/Logging/EventId.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/EventId.cs]: Removed.

File: src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs (hunk 4 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/ILogger.cs]: Removed.

File: src/Stratis.Bitcoin/Configuration/Logging/ILoggerFactory.cs (hunk 5 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/ILoggerFactory.cs]: Removed.

**Comment by quantumagi** [comment]
> @quantumagi did you test this branch with debug logging enabled? Does the other projects like Poa etc still do debug logging as they use the LogManager.GetCurrentClassLogger.

@francois, i've completed a round of testing successfully but will be re-testing due to recent changes. The `LogManager` is a new custom one.

File: src/Stratis.Bitcoin/Configuration/Logging/LogLevel.cs (hunk 6 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/LogLevel.cs]: Removed.

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs (hunk 7 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs:31]: The commented code should probably be re-introduced in some form.

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs (hunk 8 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs:48]: See above.

File: src/Stratis.Bitcoin.Features.Dns.Tests/GivenADnsFeature.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.Dns.Tests/GivenADnsFeature.cs]: Reverted.

**Comment by quantumagi** [comment]
@fassadlr , I'm still able to debug specific projects:
```
[2021-12-17 11:01:38.8124 5] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior Peer '[::ffff:66.23.237.146]:16179' connected (outbound), agent 'StratisFullNode:1.1.1.0 (80000)', height 3405336
[2021-12-17 11:01:38.8319 5] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] Sending getheaders payload with first hash: '6e30d5d9cd247ed8b9a67e70fca06915ecd15f36bc30757a8a714ffc05c4d623'.
[2021-12-17 11:01:38.8319 5] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] Sending getheaders payload with first hash: '6e30d5d9cd247ed8b9a67e70fca06915ecd15f36bc30757a8a714ffc05c4d623'.
[2021-12-17 11:01:38.8319 24] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] 0 headers were selected for sending, last one is '(null)'.
[2021-12-17 11:01:38.8319 24] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] 0 headers were selected for sending, last one is '(null)'.
[2021-12-17 11:01:39.3147 13] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] Received 0 headers. First: '(null)'  Last: '(null)'.
[2021-12-17 11:01:39.3151 13] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] Received 0 headers. First: '(null)'  Last: '(null)'.
[2021-12-17 11:01:39.3151 13] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] Headers payload with no headers was received. Assuming we're synced with the peer.
[2021-12-17 11:01:39.3151 13] DEBUG: Stratis.Bitcoin.Features.PoA.Behaviors.PoAConsensusManagerBehavior [3a06442] Headers payload with no headers was received. Assuming we're synced with the peer.
```

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs (hunk 7 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs]: Resolved.

File: src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs (hunk 8 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/Configuration/Logging/LoggingConfiguration.cs]: Resolved.

**Review by zeptin — APPROVED** [review]

**Pull request merged**